core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -1325,21 +1392,30 @@ public int getRoutingFactor() {
      * {@link org.elasticsearch.cluster.routing.OperationRouting#generateShardId(IndexMetaData, String, String)} to guarantee consistent
      * hashing / routing of documents even if the number of shards changed (ie. a shrunk index).
      *
-     * @param sourceIndexMetadata the metadata of the source index
+     * @param sourceNumberOfShards the total number of shards in the source index
      * @param targetNumberOfShards the total number of shards in the target index
      * @return the routing factor for and shrunk index with the given number of target shards.
      * @throws IllegalArgumentException if the number of source shards is less than the number of target shards or if the source shards
      * are not divisible by the number of target shards.
      */
-    public static int getRoutingFactor(IndexMetaData sourceIndexMetadata, int targetNumberOfShards) {
-        int sourceNumberOfShards = sourceIndexMetadata.getNumberOfShards();
-        if (sourceNumberOfShards < targetNumberOfShards) {
-            throw new IllegalArgumentException("the number of target shards must be less that the number of source shards");
-        }
-        int factor = sourceNumberOfShards / targetNumberOfShards;
-        if (factor * targetNumberOfShards != sourceNumberOfShards || factor <= 1) {
-            throw new IllegalArgumentException("the number of source shards [" + sourceNumberOfShards + "] must be a must be a multiple of ["
-                + targetNumberOfShards + "]");
+    public static int getRoutingFactor(int sourceNumberOfShards, int targetNumberOfShards) {
```

Comment:
I think we should remove this method. It's only used in 4 places - the setting validation (which gives us slightly odd error messages), during index creation (for validation), `selectRecoverFromShards` and `selectSplitShard`. The later two only use one part of the if in this method. I think we can inline the relevant parts of the if into those methods and have simpler code?

Reply:
why? 4 places is good?